filesystem/filesystem.go

```diff
@@ -496,10 +502,75 @@ func (m *Mount) addMetadata(path string, md metadata.Metadata, owner *user.User)
 	return m.writeDataAtomic(path, data, owner)
 }
 
+func readFull(fd int, data []byte) (int, error) {
+	i := 0
+	for i < len(data) {
+		count, err := unix.Read(fd, data[i:])
+		if err != nil {
+			return 0, err
+		}
+		if count == 0 {
+			break
+		}
+		i += count
+	}
+	return i, nil
+}
+
+// readMetadataFileSafe gets the contents of a metadata file extra-carefully,
+// considering that it could be a malicious file created to cause a
+// denial-of-service.  Specifically, the following checks are done:
+//
+// - It must be a regular file, not another type of file like a symlink or FIFO.
+// - It must have a reasonable size (<= maxMetadataFileSize).
+//
+// Take care to avoid TOCTOU (time-of-check-time-of-use) bugs when doing these
+// tests.  Notably, we must open the file before checking the file type, as the
+// file type could change between any previous checks and the open.  When doing
+// this, O_NOFOLLOW is needed to avoid following a symlink, and O_NONBLOCK is
+// needed to avoid blocking if the file is a FIFO.  Note that this implies we
+// have to use the low-level unix.Open instead of os.OpenFile.
+//
+// This function returns the data read.
+func readMetadataFileSafe(path string) ([]byte, error) {
+	fd, err := unix.Open(path, unix.O_RDONLY|unix.O_NOFOLLOW|unix.O_NONBLOCK, 0)
```

Comment:
We can use [`os.OpenFile`](https://pkg.go.dev/os#OpenFile) here as it allows passing arbitrary flags. See [the implementation here](https://cs.opensource.google/go/go/+/refs/tags/go1.17.7:src/os/file_unix.go;l=217;drc=refs%2Ftags%2Fgo1.17.7)

Note that if we do this we do not need to wrap the path error below, as the context is added by `OpenFile` IIRC.

Reply:
`O_NOFOLLOW` and `O_NONBLOCK` are in the `unix` package, not `os`.  It is unclear from the documentation that `os.OpenFile()` supports raw `unix` flags.  I suppose we can use it if it works, given that if that's true then other programs will naturally be relying on this behavior, but it would be nice to see a more definitive statement somewhere that it is supported.

Reply:
Done